File: 2_Data_Structures/C/linkedlist.c
#include "linkedlist.h"
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

struct SinglyLinkedList* InitializeNode(int value);
void PrintNode(struct Node* node); 

/* INITIALIZATION AND MEMORY FREE-ING FUNCTIONS */
struct SinglyLinkedList* InitializeLinkedList()
{
    return (struct SinglyLinkedList*)malloc(sizeof(struct SinglyLinkedList)); 
}

void FreeLinkedList(struct SinglyLinkedList* list)
{
    struct Node* current = list->head; 
    while (current != NULL) {
        struct Node* temp = current;
        current = current->next;
        printf("\nFreeing following node: "); 
        PrintNode(temp);
        free(temp);
    }
    
    printf("\nFreeing the list\n"); 
    free(list); 
}

/* QUEUE FUNCTION IMPLEMENTATIONS */

//returns 0 for success, -1 if failed (O(n) for traversal)
int AddToLinkedList(struct SinglyLinkedList* list, int value)
{
    struct Node* node = InitializeNode(value); 

    //if head is NULL, point head to the node
    if(list->head == NULL)
    {
        list->head = node; 
        return 0; 
    }

    //otherwise traverse to end of the linked list add node there
    struct Node* iterator = list->head; 

    while(iterator->next != NULL)
        iterator = iterator->next; 
    
    iterator->next = node; 
    return 0; 
 }

//returns 0 for success, -1 if failed (O(n) for traversal)
//for simplicity assumed linked in contains only distinct elements
int RemoveFromLinkedList(struct SinglyLinkedList* list, int value)
{
    //if head is NULL, return failure
    if(list->head == NULL)
    {
        return -1; 
    }

    //otherwise traverse to end of the linked list add node there
    struct Node* iterator = list->head; 

    while(iterator->next != NULL)
    {
        //if value is found - item is removed from the list, and 0 returned. 
        //as the value is already specified in argument - no need to return it. 
        if(iterator->next->value == value)
        {
            iterator->next = iterator->next->next;
            return 0; 
        }

        iterator = iterator->next;
    }

    return -1; 
} 
//returns 0 for success, non zero otherwise
int IsLinkedListEmpty(struct SinglyLinkedList* list)
{
    return list->head != NULL; 
}

/* PRINT FUNCTIONS - it's actually a traversal implementation*/
void PrintLinkedList(struct SinglyLinkedList* list)
{
    printf("\nLinkedList:");
    if(list->head == NULL)
    {
        printf("empty");
        return; 
    }
    struct Node* iterator = list->head; 
    while(iterator != NULL)
    {
        PrintNode(iterator);
        printf(" -> ");
        iterator = iterator->next; 
    }
    printf("NULL\n");
}

/*DEMO*/
void RunLinkedListInInteractiveMode(void)
{
    struct SinglyLinkedList* list = InitializeLinkedList(); 
    int result; 
    do{
        result = RunLinkedListInteractiveCycle(list); 
    }
    while (result != -1);
    
    FreeLinkedList(list); 
}

int RunLinkedListInteractiveCycle(struct SinglyLinkedList* list)
{
    printf("\n\n\nINTERACTIVE SINGLY LINKED LIST COMMANDS:\n"); 
    printf("[c] - for clearing the screen\n");
    printf("[p] - to print the list\n");
    printf("[e] - to exit\n");
    printf("[+] - to add integer to the list\n");
    printf("[-] - to remove\n");
    printf("[0] - to print if list is empty\n");
    
    char command; 
    int value; 
    int result = 0; 
    
    scanf(" %c", &command);
    printf("received %c\n", command); 
    // Process user input
        switch (command) {
            case 'c':
                system("cls"); 
                break;
            case 'p':
                PrintLinkedList(list); 
                break;
            case 'e':
                return -1; 
                break;
            case '+':
                printf("\nEnter the integer to add to the list: ");
                int result = scanf("%d", &value); 
                if(result != 0)
                {
                    printf("Entered text cannot be parsed to an integer\n");
                    break;
                }
                result = AddToLinkedList(list, value); 
                if(result != 0)
                {
                    printf("Failed to add\n");
                }
                break;
            case '-':
                printf("\nEnter the integer to remove from the list: ");
                int result = scanf("%d", &value); 
                if(result != 0)
                {
                    printf("Entered text cannot be parsed to an integer\n");
                    break;
                }
                result = RemoveFromLinkedList(list, value); 
                if(result != 0)
                {
                    printf("Failed to remove\n");
                }
                break;
            case '0':
                int isListEmpty = IsLinkedListEmpty(list);
                printf("List empty flag: %d\n", isListEmpty);
                break;
            default:
                printf("Invalid command\n");
                system("cls"); 
                break; 
        }

    return result; 
}

/* "PRIVATE" functions */
struct SinglyLinkedList* InitializeNode(int value)
{
    struct Node* node = (struct Node*)malloc(sizeof(struct Node)); 
    node->next = NULL;
    node->value = value; 

    return node; 
}

void PrintNode(struct Node* node)
{
    printf("{value: %d, address: %p}", node->value, (void*)node); 
}

File: 2_Data_Structures/C#/LinkedList.cs
using System; 
using System.Text; 

namespace DataStructures.LinkedLists; 

public class Node<T> where T : struct   
{
    public Node(T value) => Value = value; 
    public T Value { get; private set; }
    public Node<T> Next { get; private set; }
    public void SetNext(Node<T> next) => Next = next; 
    public override string ToString() => $"{Value}";
}

//For simplicity assuming linked list will contain unique elements only... 
public class LinkedList<T> where T : struct
{
    public Node<T> Head { get; private set;} 
    
    public bool Contains(T value, out Node<T> last)
    {
        last = null; 

        if(Head == null)
            return false; 

        if(Head.Value.Equals(value))
        {
            Console.WriteLine($"Head value {Head.Value} equals value {value}"); 
            return true; 
        }

        last = Head;

        while(last.Next != null)
        {
            if(last.Next.Value.Equals(value))
            {
                Console.WriteLine($"Last next value {last.Next.Value} equals value {value}");
                return true;  
            }
            
            last = last.Next; 
        }
        return false;
    }
    
    public bool Add(T value)
    {
        if(Contains(value, out var last))
            return false; 

        if(last == null)
        {
            Head = new Node<T>(value); 
            return true; 
        }
        
        last.SetNext(new Node<T>(value));  
        return true; 
    }

    public bool Remove(T value)
    {
        if(!Contains(value, out var last))
        {
            return false; 
        }

        if(last == null)
        {
            Head = Head.Next; 
            return true; 
        }

        last.SetNext(last.Next.Next); 
        return true; 
    }

    public bool IsEmpty() => Head != null; 

    public override string ToString()
    {
        var builder = new StringBuilder(); 
        builder.AppendLine("Linked List: "); 
        var temp = Head; 
        while(temp != null)
        {
            builder.Append($"{temp.ToString()} -> ");
            temp = temp.Next; 
        }
        builder.Append("NULL"); 

        return builder.ToString(); 
    } 
}

public class LinkedListRunner
{
    public void RunLinkedListInInteractiveMode()
    {
        var list = new LinkedList<int>(); 
        int result;
        Console.Clear(); 
        do {
             
            result = RunLinkedListCycle(list); 
        }
        while(result != -1); 
    }

    private int RunLinkedListCycle(LinkedList<int> list)
    {
        Console.WriteLine("\n\n\nINTERACTIVE LINKED LIST COMMANDS:\n"); 
        Console.WriteLine("[c] - for clearing the screen\n");
        Console.WriteLine("[r] - to print the linked list\n");
        Console.WriteLine("[e] - to exit\n");
        Console.WriteLine("[+] - to add an integer to the linked list\n");
        Console.WriteLine("[-] - to remove an integer from the linked list\n");
        Console.WriteLine("[?] - to check if the linked list contains an integer\n");
        char ch = Console.ReadKey().KeyChar;
        Console.WriteLine(); 

        if(ch == 'c')
        {
            Console.Clear(); 
            return 0; 
        }

        if(ch == 'r')
        {
            Console.WriteLine(list.ToString());
            return 0; 
        }

        if(ch == 'e')
            return -1; 

        if(ch == '+')
        {
            Console.WriteLine("\nEnter an integer to add to the linked list: ");
            var resultStr = Console.ReadLine().ToString(); 
            int result = Int32.Parse(resultStr); 

            var addResult = list.Add(result); 

            if(!addResult)
                Console.WriteLine("\nAdd failed, list already contains the element\n"); 

            return 0; 
        }

        if(ch == '-')
        {
            Console.WriteLine("\nEnter an integer to remove from the linked list: ");
            var resultStr = Console.ReadLine().ToString(); 
            int result = Int32.Parse(resultStr); 

            var removeResult = list.Remove(result); 

            if(!removeResult)
                Console.WriteLine("\nRemove failed, list doesn't contain the element\n"); 

            return 0; 
        }

        if(ch == '?')
        {
            Console.WriteLine("\nEnter an integer to check if linked list contains it: ");
            var resultStr = Console.ReadLine().ToString(); 
            int result = Int32.Parse(resultStr); 

            var checkResult = list.Contains(result, out _); 

            if(checkResult)
                Console.WriteLine($"\nList contains {result}.\n"); 
            else
                Console.WriteLine($"\nList doesn't contain {result}.\n"); 

            return 0; 
        }
        
        return 0;    
    }
}